File: methods/survival/templates/R/cox_ph.R
#!/usr/bin/env Rscript
# Template: Cox PH for a continuous ICI biomarker (survival / survminer).
# Primary analysis is the continuous marker (HR per 1 SD). See playbook.md §§3–5.

suppressPackageStartupMessages({
  library(survival)
})

args <- commandArgs(trailingOnly = TRUE)
if (length(args) < 1) {
  stop("usage: cox_ph.R input.csv [time_col event_col marker_col min_epv]")
}
infile <- args[[1]]
time_col <- if (length(args) >= 2) args[[2]] else "time"
event_col <- if (length(args) >= 3) args[[3]] else "event"
marker_col <- if (length(args) >= 4) args[[4]] else "marker"
min_epv <- if (length(args) >= 5) as.numeric(args[[5]]) else 10

df <- read.csv(infile, stringsAsFactors = FALSE)
df <- df[complete.cases(df[, c(time_col, event_col, marker_col)]), ]
n_events <- sum(df[[event_col]] == 1)
epv <- n_events / 1
if (epv < min_epv) {
  stop(sprintf(
    "refuse: events per variable = %.1f (%d events) < %.1f",
    epv, n_events, min_epv
  ))
}

df$z <- as.numeric(scale(df[[marker_col]]))
surv <- Surv(df[[time_col]], df[[event_col]])
fit <- coxph(surv ~ z, data = df, ties = "efron")
print(summary(fit))
cat(sprintf("Harrell C (apparent) = %.3f\n", concordance(fit)$concordance))
cat(sprintf("events per variable = %.1f\n", epv))

zph <- cox.zph(fit)
print(zph)
if (any(zph$table[, "p"] < 0.05, na.rm = TRUE)) {
  cat("PH is questionable. Report RMST or a time-varying coefficient.\n")
}

if (requireNamespace("survminer", quietly = TRUE)) {
  print(survminer::ggcoxzph(zph))
}

File: methods/survival/templates/python/cox_ph.py
#!/usr/bin/env python3
"""Template: Cox PH for a continuous ICI biomarker (lifelines).

Primary analysis is the continuous marker (HR per 1 SD). Dichotomisation is a
sensitivity analysis, not the primary model. See playbook.md sections 3–5.

Required columns: time, event (1 = event, 0 = censored), marker.
Optional: a small prespecified covariate list. Refuse the model when events
per variable fall below 10 (Peduzzi / Vittinghoff rule of thumb).
"""

from __future__ import annotations

import argparse
import sys

import pandas as pd
from lifelines import CoxPHFitter
from lifelines.statistics import proportional_hazard_test


def main() -> None:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--input", required=True, help="CSV with time, event, marker")
    ap.add_argument("--time-col", default="time")
    ap.add_argument("--event-col", default="event")
    ap.add_argument("--marker-col", default="marker")
    ap.add_argument("--covariates", default="", help="comma-separated extra columns")
    ap.add_argument("--min-epv", type=float, default=10.0)
    args = ap.parse_args()

    covs = [c.strip() for c in args.covariates.split(",") if c.strip()]
    cols = [args.time_col, args.event_col, args.marker_col] + covs
    df = pd.read_csv(args.input)[cols].dropna()
    n_events = int(df[args.event_col].sum())
    n_var = 1 + len(covs)
    epv = n_events / n_var
    if epv < args.min_epv:
        sys.exit(
            f"refuse: events per variable = {epv:.1f} "
            f"({n_events} events / {n_var} variables) < {args.min_epv}. "
            "Fit a univariable model or collect more events."
        )

    df = df.copy()
    df[args.marker_col] = (
        df[args.marker_col] - df[args.marker_col].mean()
    ) / df[args.marker_col].std(ddof=1)

    cph = CoxPHFitter()
    cph.fit(df, duration_col=args.time_col, event_col=args.event_col)
    print(cph.summary.to_string())
    print(f"\nHarrell C (apparent) = {cph.concordance_index_:.3f}")
    print(f"events per variable = {epv:.1f}")

    zph = proportional_hazard_test(cph, df, time_transform=["km", "rank"])
    print("\nSchoenfeld PH tests:")
    print(zph.summary.to_string())
    if (zph.summary["p"] < 0.05).any():
        print(
            "PH is questionable. Report RMST or a time-varying coefficient; "
            "do not quote a single HR as if it were constant."
        )


if __name__ == "__main__":
    main()
